python/lsst/pipe/base/execution_reports.py

```diff
@@ -0,0 +1,438 @@
+                    break
+            else:
+                self.missing_failed.add(output_ref)
+        else:
+            status_graph.nodes[output_ref.id]["not_produced"] = True
+            self.missing_not_produced[output_ref] = any(
+                status_graph.nodes[upstream_dataset_id].get("not_produced", False)
+                for upstream_quantum_id in status_graph.predecessors(output_ref.id)
+                for upstream_dataset_id in status_graph.predecessors(upstream_quantum_id)
+            )
+
+    def handle_produced_dataset(self, output_ref: DatasetRef, status_graph: networkx.DiGraph) -> None:
+        """Account for produced datasets.
+
+        Parameters
+        ----------
+        output_ref: `DatasetRef`
+            Dataset reference of the dataset.
+        status_graph: `networkx.DiGraph`
+            The quantum graph produced by
+            QuantumGraphExecutionReport.make_reports() which steps through the
+            quantum graph of a run and logs the status of each quantum.
+
+        See Also
+        --------
+        TaskExecutionReport.inspect_quantum
+        """
+        status_graph.nodes[output_ref.id]["not_produced"] = False
+        self.n_produced += 1
+
+
+@dataclasses.dataclass
+class TaskExecutionReport:
+    """A report on the status and content of a task in an executed quantum
+    graph.
+
+    Use task metadata to identify and inspect failures and report on output
+    datasets.
+
+    See Also
+    --------
+    QuantumGraphExecutionReport
+    DatasetTypeExecutionReport
+    """
+
+    failed: dict[uuid.UUID, DatasetRef] = dataclasses.field(default_factory=dict)
+    """A mapping from quantum data ID to log dataset reference for quanta that
+    failed directly in this run (`dict`).
+    """
+    failed_upstream: dict[uuid.UUID, DataCoordinate] = dataclasses.field(default_factory=dict)
+    """A mapping of data IDs of quanta that were not attempted due to an
+    upstream failure (`dict`).
+    """
+    output_datasets: dict[str, DatasetTypeExecutionReport] = dataclasses.field(default_factory=dict)
+    """Missing and produced outputs of each DatasetType (`dict`).
+    """
+
+    def inspect_quantum(
+        self,
+        quantum_node: QuantumNode,
+        status_graph: networkx.DiGraph,
+        refs: Mapping[str, Mapping[uuid.UUID, DatasetRef]],
+        metadata_name: str,
+        log_name: str,
+    ) -> None:
+        """Inspect a quantum of a quantum graph and ascertain the status of
+        each associated data product.
+
+        Parameters
+        ----------
+        quantum_node: `QuantumNode`
+            The specific node of the quantum graph to be inspected.
+        status_graph: `networkx.DiGraph`
+            The quantum graph produced by
+            QuantumGraphExecutionReport.make_reports which steps through the
+            quantum graph of a run and logs the status of each quantum.
+        refs: `Mapping[str, Mapping[uuid.UUID, DatasetRef]]`
```

Comment:
```suggestion
        refs: `~collections.abc.Mapping` [ `str`, `~collections.abc.Mapping` [ \
                `uuid.UUID`, `~lsst.daf.butler.DatasetRef ] ]
```
Unfortunately you can't put backticks around the whole thing; each symbol needs its own backticks.  And symbols like `Mapping` and `DatasetRef` need to be fully-qualified if they're not in `lsst.pipe.base`.